### PR TITLE
Kotlin: Revert type erasure within $default functions

### PR DESCRIPTION
**Pull request by smowton**
This imprecise implementation turned out to cause linkage errors, e.g. when type variables in the signatures of member methods were inappropriately erased. For the time being we accept that $default methods despite having erased signatures in keeping with their JVM representation can contain expressions whose types make reference to their surrounding function or class' type variables, even though they should be out of scope since $default methods are static and don't have type parameters, and need to cope with the inconsistency in QL.

### CODE REVIEW

File: java/ql/test/kotlin/library-tests/parameter-defaults/erasure.ql [hunk 1]
```diff
@@ -5,6 +5,8 @@ class InstantiatedType extends ParameterizedType {
 }
 
 // This checks that all type parameter references are erased in the context of a $default function.
+// Note this is currently expected to fail since for the time being we extract type variable references
+// even where they should be out of scope.
```
Comment (tamasvajk) [inline java/ql/test/kotlin/library-tests/parameter-defaults/erasure.ql:9]: I don't think this is specific to `$default` functions. This applies in general: type variables should not be referenced if they are not in scope. Should this test be turned into a consistency query?

Comment (smowton) [reply java/ql/test/kotlin/library-tests/parameter-defaults/erasure.ql:9]: Yes, I just haven't done the work yet :)

**Comment by smowton** [comment]
Have confirmed that with this PR the only remaining db inconsistencies in an extraction of kotlin-1.5.30 relate to some genuinely doubly-compiled files producing duplicate exprs and stmts since they always get fresh names

**Review by tamasvajk — APPROVED** [review]
LGTM. Sorry for not approving before, I thought you'd handle https://github.com/github/codeql/pull/11665#discussion_r1046869820 in this PR.

**Pull request merged**